src/direct_copy_stubs.c

```diff
@@ -0,0 +1,137 @@
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ */
+
+#define _GNU_SOURCE
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/ioctl.h>
+
+#include <errno.h>
+#include <fcntl.h>
+#include <string.h>
+#include <unistd.h>
+#include <stdint.h>
+
+#include <caml/alloc.h>
+#include <caml/memory.h>
+#include <caml/signals.h>
+#include <caml/fail.h>
+#include <caml/callback.h>
+#include <caml/bigarray.h>
+
+#ifdef __linux__
+# include <linux/fs.h>
+# include <sys/sendfile.h>
+# include <malloc.h>
+#endif
+
+/* ocaml/ocaml/unixsupport.c */
+extern void uerror(char *cmdname, value cmdarg);
+#define Nothing ((value) 0)
+
+#define TRIED_AND_FAILED (1)
+#define OK 0
+
+#define XFER_BUFSIZ (2*1024*1024)
+
+struct direct_copy_handle {
+  int  in_fd;
+  int  out_fd;
+  char *buffer;
+};
+
+CAMLprim value stub_init(value in_fd, value out_fd)
+{
+  CAMLparam2(in_fd, out_fd);
+  int c_in_fd = Int_val(in_fd);
+  int c_out_fd = Int_val(out_fd);
+  struct direct_copy_handle *cpinfo = NULL;
+  int flags;
+
+  /* initialise handle */
+  cpinfo = malloc(sizeof(struct direct_copy_handle));
+  if (!cpinfo) caml_raise_out_of_memory();
+  cpinfo->buffer = NULL;
+  if (posix_memalign((void **)&cpinfo->buffer, sysconf(_SC_PAGESIZE), XFER_BUFSIZ)) {
+      free(cpinfo);
+      caml_raise_out_of_memory();
+  }
+  cpinfo->in_fd = c_in_fd;
+  cpinfo->out_fd = c_out_fd;
+
+  /* Force the output to have O_DIRECT if possible */
+  flags = fcntl(c_out_fd, F_GETFL, NULL);
+  if (flags > 0 && !(flags & O_DIRECT))
+    fcntl(c_out_fd, F_SETFL, flags | O_DIRECT);
+
+  CAMLreturn((value) cpinfo);
+
+}
+
+CAMLprim value stub_cleanup(value handle)
+{
+  CAMLparam1(handle);
+  struct direct_copy_handle *cpinfo = (struct direct_copy_handle *)handle;
+
+  free(cpinfo->buffer);
+  free(cpinfo);
+  CAMLreturn(Val_unit);
+}
+
+CAMLprim value stub_direct_copy(value handle, value len){
+  CAMLparam2(handle, len);
+  CAMLlocal1(result);
+  size_t c_len = Int64_val(len);
+  struct direct_copy_handle *cpinfo = (struct direct_copy_handle *)handle;
+  size_t bytes;
+  size_t remaining;
+  int    rc;
+
+  enter_blocking_section();
+
+  rc = TRIED_AND_FAILED;
+  bytes = 0;
+  
+  remaining = c_len;
+  while (remaining > 0) {
+    ssize_t bread;
+    ssize_t bwritten = 0;
+
+    bread = read(cpinfo->in_fd, cpinfo->buffer, (remaining < XFER_BUFSIZ)?remaining:XFER_BUFSIZ) ;
+    if (bread < 0) goto fail;
+    while (bwritten < bread) {
+      ssize_t ret;
+
+      ret = write(cpinfo->out_fd, cpinfo->buffer + bwritten, bread - bwritten);
```

Comment:
However unlike getting a 0 return from read(), getting a 0 return from write() is an unexpected EOF for us. I think I need to add an RC for that because my existing uerror() won't do anything useful.